packages/server/src/data/votes-data.js

```diff
@@ -0,0 +1,163 @@
+const _ = require('lodash');
+const assert = require('assert');
+
+const { DataBase } = require('./data-base');
+const { VoteModel } = require('../models');
+
+const { Params } = require('./classes');
+const { Query } = require('./classes');
+const { Options } = require('./classes');
+const { Data } = require('./classes');
+
+const logger = require('../lib/log');
+
+/**
+ * @implements {Layers.DataSourceLayer.DataSource}
+ * @property {VoteModel} model
+ */
+class VotesData extends DataBase {
+  constructor() {
+    super();
+    this.model = new VoteModel();
+    this.log = logger.module('VotesData');
+  }
+
+  /**
+   * @param {Models.MeasurementID} measurementId
+   * @returns {Promise<number | null>}
+   * @public
+   */
+  async totalVoteAmount(measurementId) {
+    const params = new Params({ measurementId });
+    const query = new Query({ sort: 0 });
+    const options = new Options({ raw: true, attributes: true, group: true });
+    const votes = await this.getOne(params, query, options);
+    return _.get(votes, 'totalVoteAmount', null);
+  }
+
+  /**
+   * @param {Models.AgentID} agentId
+   * @param {Models.MeasurementID} measurementId
+   * @returns {Promise<Models.Vote>}
+   * @public
+   */
+  async upvote(agentId, measurementId) {
+    return this._changeVote(agentId, measurementId, this._up);
+  }
+
+  /**
+   * @param {Models.AgentID} agentId
+   * @param {Models.MeasurementID} measurementId
+   * @returns {Promise<Models.Vote>}
+   * @public
+   */
+  async downvote(agentId, measurementId) {
+    return this._changeVote(agentId, measurementId, this._down);
```

Comment:
From what I can tell, the client is expected to send ``upvote`` and ``downvote`` requests. Each one increments or decrements the votes by 1. 

I think this makes me a bit uncomfortable. I could imagine that voting will lag a bit; so when a user clicks "Upvote" 3 times, it will take a while for each to reach the backend.

What would you think of this approach:
The client instead sends a total vote count, like "-4" or "+4". This is sent to the server when they stop clicking (there's a short debouncing).

Reply:
Ok, will fix it.